tfx/components/example_gen/utils.py

```diff
@@ -209,84 +213,180 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
+    # is ['{SPAN}', '{VERSION}'], since a newer span dominates a newer version.
+    properties = [result.group(name) for name in property_names]
+    update_rest = False
+
+    # Uses str instead of int because of zero padding digits.
+    for i, prop_str in enumerate(properties):
+      try:
+        prop = int(prop_str)
+      except ValueError:
+        raise ValueError('Cannot find %s number from %s based on %s' %
+                         (property_names[i], file_path, split_regex_pattern))
+
+      # Since specs are in descreasing order of performance, if we find
+      # a newer property value, all subsequently less important properties
+      # must be updated to align with that new property.
+      if update_rest:
+        latest_properties[i] = prop_str
+      elif latest_properties[i] is None or prop >= int(latest_properties[i]):
+        latest_properties[i] = prop_str
+        update_rest = True
+
+  if any(prop is None for prop in latest_properties):
+    raise ValueError('Cannot find matching for split %s based on %s' %
                      (split.name, split.pattern))
-  return latest_span
+
+  return latest_properties
+
+
+def _retrieve_latest_span_version(uri: Text,
+                                  split: example_gen_pb2.Input.Split
+                                  ) -> Tuple[Optional[Text], Optional[Text]]:
+  """Retrieves the most recent span and version for a given split pattern.
+
+  If both Span and Version spec occur in the split pattern, searches for and
+  returns both the latest Span and Version. If only Span exists in the split
+  pattern, searches for the latest Span, and Version is returned as None.
+  If Version is present, but not Span, an error is raised. If neither Span
+  nor Version is present, returns both as None.
+
+  Args:
+    uri: The base path from which files will be searched.
+    split: An example_gen_pb2.Input.Split object which contains a split pattern,
+      to be searched on. Note that this function will update the {SPAN} in this
+      and {VERSION} tags in the split config to actual Span and Version numbers.
+  Returns:
+    Tuple of two strings, Span (optional) and Version (optional).
+
+  Raises:
+    ValueError: if any of the following occurs:
+      - If either Span or Version spec is occurs in the split pattern
+        more than once.
+      - If Version spec is provided, but Span spec is not present.
+      - If Span or Version found is not an integer.
+      - If a matching cannot be found for split pattern provided.
+  """
+
+  latest_span = None
+  latest_version = None
+
+  if SPAN_SPEC in split.pattern:
+    if split.pattern.count(SPAN_SPEC) != 1:
+      raise ValueError('Only one {SPAN} is allowed in %s' % split_pattern)
+
+    if VERSION_SPEC in split.pattern:
+      if split.pattern.count(VERSION_SPEC) != 1:
+        raise ValueError('Only one {VERSION} is allowed in %s' % split_pattern)
+
+      latest_span, latest_version = _property_search(uri, split,
```

Comment:
nit, rename _property_search to `_glob_and_match_span_version`, which may read better

Reply:
fixed